src/d3d9/d3d9_mem.cpp

```diff
@@ -0,0 +1,350 @@
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    m_allocatedMemory -= Chunk->Size();
+
+    m_chunks.erase(std::remove_if(m_chunks.begin(), m_chunks.end(), [&](auto& item) {
+        return item.get() == Chunk;
+    }), m_chunks.end());
+  }
+
+  void D3D9MemoryAllocator::NotifyMapped(uint32_t Size) {
+    m_mappedMemory += Size;
+  }
+
+  void D3D9MemoryAllocator::NotifyUnmapped(uint32_t Size) {
+    m_mappedMemory -= Size;
+  }
+
+  void D3D9MemoryAllocator::NotifyFreed(uint32_t Size) {
+    m_usedMemory -= Size;
+  }
+
+  uint32_t D3D9MemoryAllocator::MappedMemory() {
+    return m_mappedMemory.load();
+  }
+
+  uint32_t D3D9MemoryAllocator::UsedMemory() {
+    return m_usedMemory.load();
+  }
+
+  uint32_t D3D9MemoryAllocator::AllocatedMemory() {
+    return m_allocatedMemory.load();
+  }
+
+  D3D9MemoryChunk::D3D9MemoryChunk(D3D9MemoryAllocator* Allocator, uint32_t Size)
+    : m_allocator(Allocator), m_size(Size), m_mappingGranularity(m_allocator->MemoryGranularity() * 16) {
+    m_mapping = CreateFileMappingA(INVALID_HANDLE_VALUE, nullptr, PAGE_READWRITE | SEC_COMMIT, 0, Size, nullptr);
+    m_freeRanges.push_back({ 0, Size });
+    m_mappingRanges.resize(((Size + m_mappingGranularity - 1) / m_mappingGranularity));
+  }
+
+  D3D9MemoryChunk::~D3D9MemoryChunk() {
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    CloseHandle(m_mapping);
+  }
+
+  void* D3D9MemoryChunk::Map(D3D9Memory* memory) {
+    std::lock_guard<dxvk::mutex> lock(m_mutex);
+
+    uint32_t alignedOffset = alignDown(memory->GetOffset(), m_mappingGranularity);
+    uint32_t alignmentDelta = memory->GetOffset() - alignedOffset;
+    uint32_t alignedSize = memory->GetSize() + alignmentDelta;
+    if (alignedSize > m_mappingGranularity) {
+      // The allocation crosses the boundary of the internal mapping page it's a part of
+      // so we map it on it's own.
+      alignedOffset = alignDown(memory->GetOffset(), m_allocator->MemoryGranularity());
+      alignmentDelta = memory->GetOffset() - alignedOffset;
+      alignedSize = memory->GetSize() + alignmentDelta;
+
+      m_allocator->NotifyMapped(alignedSize);
+      uint8_t* basePtr = static_cast<uint8_t*>(MapViewOfFile(m_mapping, FILE_MAP_ALL_ACCESS, 0, alignedOffset, alignedSize));
+      if (unlikely(basePtr == nullptr)) {
+        DWORD error = GetLastError();
+        LPTSTR buffer = nullptr;
+        FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM, nullptr, error, MAKELANGID(LANG_NEUTRAL, SUBLANG_NEUTRAL), (LPTSTR)&buffer, 0, nullptr);
+        Logger::err(str::format("Mapping non-persisted file failed: ", error, ", Mapped memory: ", m_allocator->MappedMemory(), ", Msg: ", buffer));
+        if (buffer) {
+          LocalFree(buffer);
+        }
+        return nullptr;
+      }
+      return basePtr + alignmentDelta;
+    }
+
+    // For small allocations we map the entire mapping page to minimize the overhead from having the align the offset to 65k bytes.
```

Comment:
Does the allocator make sure that small allocations can't straddle two pages?

Reply:
Yes, the if just before this line takes care of that.

Reply:
Right, that's what all the alignDown mess etc is for.

Is there any downside to having the same page included in multiple mappings, or having overlapped mappings? Both can still happen with the current logic.

Reply:
The docs aren't super detailled except for saying that it's coherent. There's no mention of any downsides there and it works fine in practice, so I assume there isnt.